apps/builder/app/builder/features/ai/ai-fetch-result.ts

```diff
@@ -33,38 +33,53 @@ import { restAi } from "~/shared/router-utils";
 import untruncateJson from "untruncate-json";
 import { traverseTemplate } from "@webstudio-is/jsx-utils";
 import { RequestParamsSchema } from "~/routes/rest.ai._index";
+import {
+  AiApiException,
+  RateLimitException,
+  textToRateLimitMeta,
+} from "./api-exceptions";
 
 const unknownArray = z.array(z.unknown());
 
+const onResponseReceived = async (response: Response) => {
+  if (response.ok === false) {
+    const text = await response.text();
+
+    if (response.status === 429) {
+      const meta = textToRateLimitMeta(text);
+      throw new RateLimitException(text, meta);
+    }
+
+    throw new Error(
+      `Fetch error status="${response.status}" text="${text.slice(0, 1000)}"`
+    );
+  }
+};
+
 export const fetchResult = async (
   prompt: string,
   abortSignal: AbortSignal
-): Promise<string[]> => {
+): Promise<void> => {
   const commandsResponse = await handleAiRequest<commandDetect.Response>(
     fetch(restAi("detect"), {
       method: "POST",
       body: JSON.stringify({ prompt }),
       signal: abortSignal,
     }),
-    { signal: abortSignal }
-  );
-
-  if (
-    commandsResponse.success === false ||
-    commandsResponse.type === "stream"
-  ) {
-    // Commands detection is not using streaming
-    // therefore this should never be the case.
-    // The check is here just for type safety.
-    if (commandsResponse.type === "stream") {
-      return ["Something went wrong."];
+    {
+      onResponseReceived,
     }
+  );
 
-    if (abortSignal.aborted === false) {
-      return [commandsResponse.data.message];
-    }
+  if (commandsResponse.type === "stream") {
+    throw new Error(
+      "Commands detection is not using streaming. Something went wrong."
```

Comment:
Hmm is this going to be user-facing?

Reply:
Not this PR change. Should never throw, but in case if it is it will be visible to user. Same for all kinds of unexpected errors.